### PR TITLE
Improve S5443 performance: Reuse compiled Regex

### PR DESCRIPTION
**Pull request by martin-strecker-sonarsource**
see also #8181
Strips of another 4,5 seconds (or 0,08%) by caching the JITed regex in `IsSensitiveDirectoryUsage`:

### JIT

Before:
image

After:
image

### Overall

Before
image

After
image

### CODE REVIEW

**Comment by martin-strecker-sonarsource** [comment]
Regarding the code smells/security hotspots:
* https://github.com/SonarSource/sonar-dotnet/pull/8185#discussion_r1381623489
* https://github.com/SonarSource/sonar-dotnet/pull/8185#discussion_r1381797345
* https://github.com/SonarSource/sonar-dotnet/pull/8181#discussion_r1381895146

File: analyzers/src/SonarAnalyzer.Common/Rules/Hotspots/PubliclyWritableDirectoriesBase.cs [hunk 1]
```diff
@@ -102,11 +96,29 @@ protected override void Initialize(SonarAnalysisContext context)
...
+        private static string[] MacDirs() => new[]
+            {
+                "/var/tmp",
+                "/usr/tmp",
+                "/dev/shm",
+                "/library/caches",
+                "/users/shared",
+                "/private/tmp",
+                "/private/var/tmp",
+            };
```
Comment (sebastien-marichal) [inline analyzers/src/SonarAnalyzer.Common/Rules/Hotspots/PubliclyWritableDirectoriesBase.cs:122]: Why switch from fields to methods?
What are the benefits of not making the fields static?

Comment (martin-strecker-sonarsource) [reply analyzers/src/SonarAnalyzer.Common/Rules/Hotspots/PubliclyWritableDirectoriesBase.cs:122]: They are only ever used in the static ctor. There is no need to keep them in a static field afterward.

File: analyzers/src/SonarAnalyzer.Common/Rules/Hotspots/PubliclyWritableDirectoriesBase.cs [hunk 2]
```diff
@@ -29,27 +29,26 @@ public abstract class PubliclyWritableDirectoriesBase<TSyntaxKind, TInvocationEx
...
-            "/private/tmp",
-            "/private/var/tmp",
-        };
-        private readonly Regex userProfile = new(@"^%USERPROFILE%[\\\/]AppData[\\\/]Local[\\\/]Temp", WindowsAndUnixOptions);
-        private readonly Regex linuxDirectories;
-        private readonly Regex macDirectories;
-        private readonly Regex windowsDirectories = new(@"^([a-z]:[\\\/]?|[\\\/][\\\/][^\\\/]+[\\\/]|[\\\/])(windows[\\\/])?te?mp([\\\/]|$)", WindowsAndUnixOptions);
-        private readonly Regex environmentVariables;
+            var insecureEnvironmentVariables = new[] { "tmp", "temp", "tmpdir" };
+            InsecureEnvironmentVariables = insecureEnvironmentVariables;
```
Comment (sebastien-marichal) [inline analyzers/src/SonarAnalyzer.Common/Rules/Hotspots/PubliclyWritableDirectoriesBase.cs:44]: Why are you using an intermediate variable for `InsecureEnvironmentVariables`?
```suggestion
            InsecureEnvironmentVariables = new[] { "tmp", "temp", "tmpdir" };
```

File: analyzers/src/SonarAnalyzer.Common/Rules/Hotspots/PubliclyWritableDirectoriesBase.cs [hunk 3]
```diff
@@ -29,27 +29,26 @@ public abstract class PubliclyWritableDirectoriesBase<TSyntaxKind, TInvocationEx
...
+            UserProfile = new("""^%USERPROFILE%[\\\/]AppData[\\\/]Local[\\\/]Temp""", WindowsAndUnixOptions);
+            LinuxDirectories = new($@"^({LinuxDirs().JoinStr("|", Regex.Escape)})(\/|$)", RegexOptions.Compiled);
```
Comment (pavel-mikula-sonarsource) [inline analyzers/src/SonarAnalyzer.Common/Rules/Hotspots/PubliclyWritableDirectoriesBase.cs:46]: I would question the use of `RegexOptions.Compile` in general in this rule. The cost of creating in memory assembly with emitting instructions at runtime to evaluate regex is extremely high, often not worth the benefit of the speed bump. 

These regexes are generally straight-forward with not so much backtracking.

Did you try to measure the impact of the `Compiled` thing?

File: analyzers/src/SonarAnalyzer.Common/Rules/Hotspots/PubliclyWritableDirectoriesBase.cs (hunk 2 above)
Comment (martin-strecker-sonarsource) [reply analyzers/src/SonarAnalyzer.Common/Rules/Hotspots/PubliclyWritableDirectoriesBase.cs:44]: That way, the property assignments can be safely re-ordered.

File: analyzers/src/SonarAnalyzer.Common/Rules/Hotspots/PubliclyWritableDirectoriesBase.cs (hunk 3 above)
Comment (martin-strecker-sonarsource) [reply analyzers/src/SonarAnalyzer.Common/Rules/Hotspots/PubliclyWritableDirectoriesBase.cs:46]: This is not what the PR is about. The change shows significant improvements. Whether "Compiled" is bringing gains needs to be decided for all Regexes and a wide variety of projects. Create an issue, if you want us to look into it.

**Review by sebastien-marichal — APPROVED** [review]
LGTM!

**Pull request merged**